src/xml2_url.cpp

```diff
@@ -1,156 +1,133 @@
-    INTEGER(port)[i] = uri->port == 0 ? NA_INTEGER : uri->port;
-    SET_STRING_ELT(user, i, Rf_mkChar(uri->user == NULL ? "" : uri->user));
-    SET_STRING_ELT(path, i, Rf_mkChar(uri->path  == NULL ? "" : uri->path));
-    SET_STRING_ELT(fragment, i, Rf_mkChar(uri->fragment == NULL ? "" : uri->fragment));
+    scheme[i] = uri->scheme == NULL ? "" : uri->scheme;
+    server[i] = uri->server == NULL ? "" : uri->server;
+    port[i] = uri->port == 0 ? NA_INTEGER : uri->port;
+    user[i] = uri->user == NULL ? "" : uri->user;
+    path[i] = uri->path  == NULL ? "" : uri->path;
+    fragment[i] = uri->fragment == NULL ? "" : uri->fragment;
 
     /* * *
      * Thu Apr 26 10:36:26 CEST 2007 Daniel Veillard
      * svn path=/trunk/; revision=3607
      * https://github.com/GNOME/libxml2/commit/a1413b84f7163d57c6251d5f4251186368efd859
      */
     #if defined(LIBXML_VERSION) && (LIBXML_VERSION >= 20629)
-    SET_STRING_ELT(query, i, Rf_mkChar(uri->query_raw  == NULL ? "" : uri->query_raw));
+    query[i] = uri->query_raw  == NULL ? "" : uri->query_raw;
     #else
-    SET_STRING_ELT(query, i, Rf_mkChar(uri->query  == NULL ? "" : uri->query));
+    query[i] = uri->query  == NULL ? "" : uri->query;
     #endif
 
     xmlFreeURI(uri);
   }
 
-  SEXP out = PROTECT(Rf_allocVector(VECSXP, 7));
-  SET_VECTOR_ELT(out, 0, scheme);
-  SET_VECTOR_ELT(out, 1, server);
-  SET_VECTOR_ELT(out, 2, port);
-  SET_VECTOR_ELT(out, 3, user);
-  SET_VECTOR_ELT(out, 4, path);
-  SET_VECTOR_ELT(out, 5, query);
-  SET_VECTOR_ELT(out, 6, fragment);
+  using namespace cpp11::literals;
 
-  SEXP names = PROTECT(Rf_allocVector(STRSXP, 7));
-
-  SET_STRING_ELT(names, 0, Rf_mkChar("scheme"));
-  SET_STRING_ELT(names, 1, Rf_mkChar("server"));
-  SET_STRING_ELT(names, 2, Rf_mkChar("port"));
-  SET_STRING_ELT(names, 3, Rf_mkChar("user"));
-  SET_STRING_ELT(names, 4, Rf_mkChar("path"));
-  SET_STRING_ELT(names, 5, Rf_mkChar("query"));
-  SET_STRING_ELT(names, 6, Rf_mkChar("fragment"));
-
-  Rf_setAttrib(out, R_ClassSymbol, Rf_mkString("data.frame"));
-  Rf_setAttrib(out, R_NamesSymbol, names);
-
-  SEXP row_names = PROTECT(Rf_allocVector(INTSXP, 2));
-  INTEGER(row_names)[0] = NA_INTEGER;
-  INTEGER(row_names)[1] = -n;
-  Rf_setAttrib(out, R_RowNamesSymbol, row_names);
-
-  UNPROTECT(10);
+  cpp11::writable::data_frame out({
+    "scheme"_nm = scheme,
+    "server"_nm = server,
+    "port"_nm = port,
+    "user"_nm = user,
+    "path"_nm = path,
+    "query"_nm = query,
+    "fragment"_nm = fragment,
+  });
 
   return out;
 }
 
-// [[export]]
-extern "C" SEXP url_escape_(SEXP x_sxp, SEXP reserved_sxp) {
-  R_xlen_t n = Rf_xlength(x_sxp);
-  SEXP out = PROTECT(Rf_allocVector(STRSXP, n));
-
-  if (Rf_xlength(reserved_sxp) != 1) {
-    Rf_error("`reserved` must be character vector of length 1");
-  }
+[[cpp11::register]]
+cpp11::strings url_escape_(cpp11::strings x_sxp, cpp11::strings reserved_sxp) {
+  int n = x_sxp.size();
```

Comment:
The size method does actually still return a `R_xlen_t`, so it is probably best to keep that everywhere if it is easy
https://github.com/r-lib/cpp11/blob/500f642b4ea132ec8c168fc70a28e81e9510ece3/inst/include/cpp11/r_vector.hpp#L114